aten/src/ATen/native/cuda/ForeachReduceOp.cu

```diff
@@ -134,9 +135,18 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
         return at::isIntegralType(scalar_type, /*includeBool*/ true) ||
             at::isComplexType(scalar_type);
       });
-  if (!can_use_fast_route(tensors) || has_int_or_complex ||
+  const bool is_same_dtype = dtype.has_value()
```

Comment:
check_fast_path_restrictions already makes this check (which is called from can_use_fast_route, I believe) so this check shouldn't be necessary

Reply:
> check_fast_path_restrictions already checks that all tensors are the same dtype (which is called from can_use_fast_route, I believe), so would this check against the optional argument dtype still be necessary?

Yes, essentially you can set all tensors to fp16 and specify `dtype=torch.float`. The goal is to cast first and then compute the norm similarly to: https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html#torch.linalg.vector_norm

> Or, if you're verifying that the tensors should match with the optional argument dtype, then checking just the first one will do right?

Actually this is true, the only issue with that is that you'd be using knowledge of what's in `can_use_fast`. I guess the same argument could be made to `has_int_or_complex` right above. You could compare just one assuming that `can_use_fast` already checks if all tensors have same dtype.

 > Maybe I'm a little confused--why are we checking that any of the tensors are the same dtype as the optional argument dtype?

That's bad implem for my end. It's supposed to be `all`. I've replace the implementation by computing the opposite (which should be any and removed the `!` in the if condition)